### ISSUE TITLE
Incorrect convex hull for contour.

### ISSUE THREAD

**Issue by ZhivaevDmitry** [opencv/opencv#24907]
### System Information

OpenCV python version: 4.8.1
Operating System / Platform: Windows 10
Python version: 3.11.5

### Detailed description

The "convexHull" function produces an incorrect result for the outer contour, making any further data processing and information extraction from the contour impossible. In my opinion, this contour is correct and does not have self-intersections, but it has one point of self-contact, which causes the incorrect result.

![zoom_0](https://github.com/opencv/opencv/assets/11032016/d14d7152-6701-4667-a448-bbabe622f779)
![zoom](https://github.com/opencv/opencv/assets/11032016/c0406180-1d55-4c5e-945a-20e295f11498)

Output log:
```
Contour:
[...

 [[ 74 100]]

 [[ 74 102]]

 [[ 73 103]]

 [[ 72 103]]

 [[ 73 103]]

 [[ 74 102]]

 [[ 83 102]]

 ...]
Convex hull:
[
 ...
 [37]
 [47]
 [46]
 [50]
 ...]
```

### Steps to reproduce

[greyscale.zip](https://github.com/opencv/opencv/files/14013026/greyscale.zip)

```
import cv2 as cv

src_greyscale = cv.imread("greyscale.png", cv.IMREAD_GRAYSCALE)

contours, hierarchy = cv.findContours(src_greyscale, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)  

convex_hull_indices = cv.convexHull(contours[0], clockwise = True, returnPoints = False)

print("Contour:")
print(contours[0])
print("Convex hull:")
print(convex_hull_indices)
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by mshabunin** [comment]
@ZhivaevDmitry , isn't it a duplicate of #4954?

**Comment by ZhivaevDmitry** [comment]
> @ZhivaevDmitry , isn't it a duplicate of #4954?

Yes, that's it, and for 8 years, no one bothered to fix it. I think it's time to remind about this issue again. I really hope we won't forget about it for another 8 years. Considering that it's an extremely trivial algorithm, which is very easy to implement. Finding the convex hull is just the basics. Sorry for the emotions.

**Comment by LaurentBerger** [comment]
https://math.stackexchange.com/questions/2632754/why-is-sklansky-algorithm-convex-hull-wrong

**Comment by ZhivaevDmitry** [comment]
> https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#ga014b28e56cb8854c0de4a211cb2be656 https://docs.opencv.org/4.x/d0/de3/citelist.html#CITEREF_Sklansky82 https://math.stackexchange.com/questions/2632754/why-is-sklansky-algorithm-convex-hull-wrong
> 
> sklansky-algorithm for [convex hull is wrong](https://github.com/opencv/opencv/issues/4954#issuecomment-268407733) then opencv result for convex hull is wrong

It's extremely interesting that the incorrectness of the algorithm has already been identified and proven. Perhaps some hybrid scheme will allow solving this problem, a mixture of several algorithms. This issue, as I understand it, arises extremely rarely, and I accidentally discovered it when adding random noise to my original dataset. Out of 100,000 original images, this error manifested itself in only one.

As I see it, it is quite clear when the convex hull becomes incorrect - this can be easily detected if Monotonous is violated. Perhaps it is worth adding some post-processor that would trigger in case of Monotonous violation.

**Comment by mshabunin** [comment]
I've done a bit more digging and found related issues #4539 and #5908. Perhaps the root cause of all problems is that `convexHull` function does not accept contour, but a set of points. While `convexityDefects` requres contour (simple polygon) as input. Due to this `convexHull` can shuffle points and produce incorrect output from the contour (simple polygon) convex hull point of view.

So I beleive instead of patching this mess further we need either separate function to calculate convex hull of contours (simple polygons) or a flag/mode/autodetection in the current `convexHull` function.

Additionally, looks like contours returned by `findContours` are not _simple polygons_ because they might have _points of self-contact_ which are basically self-intersections. So additional output processing is required here as well.

Links:
* https://en.wikipedia.org/wiki/Convex_hull_algorithms
* https://en.wikipedia.org/wiki/Convex_hull_of_a_simple_polygon
* Review of algorithms for calculating convex hull for simple polygon: https://cgm.cs.mcgill.ca/~athens/cs601/

**Comment by YanBC** [comment]
@ZhivaevDmitry I apologize if this sounds naive and perhaps I do not fully understand your problem. However since point `46` and point `48` are the same why does this final convex hull is considered wrong, 
```
Wrong convex hull:
[
 ...
 [37]
 [47]
 [46]
 [50]
 ...]
```
while the following convex hull is correct?
```
Correct convex hull:
[
 ...
 [37]
 [47]
 [48]
 [50]
 ...]
```

**Comment by ZhivaevDmitry** [comment]
@YanBC Somewhere between versions **4.1** and **4.9** of **OpenCV**, the code responsible for **cv.convexHull** and **cv.convexityDefects** was modified. According to the new internal rules, **cv.convexHull** is supposed to (but doesn't always) return a monotonous array of convex hull indices, i.e., **h[i] < h[i+1]**. Furthermore, now if you pass a non-monotonous array of convex hull indices to **cv.convexityDefects**, an error will be thrown:
> .../convhull.cpp:360: error: (-5 argument) The convex hull indices are not monotonous, which can be in the case when the input contour contains self-intersections in function 'convexityDefects'

**Comment by opencv-alalek** [comment]
relates #16594

**Comment by ZhivaevDmitry** [comment]
@opencv-alalek Yes, here is another minimalist example demonstrating the problem:

**input:** ![grayscale](https://github.com/user-attachments/assets/3675a1a0-8c22-46ee-8d4e-e922b9e0e896)

```
black_and_white = cv.imread(path_to_file, cv.IMREAD_GRAYSCALE)

print (black_and_white)
    
contours, hierarchy = cv.findContours(black_and_white, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)        

convex_hull_indices = cv.convexHull(contours[0], clockwise = True, returnPoints = False)

print(convex_hull_indices)
```

results:
```
[[  0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0]
 [  0   0   0 255 255 255 255   0]
 [  0   0   0 255 255 255 255   0]
 [  0   0   0 255 255 255 255   0]
 [  0 255 255   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0]]
 

[[3]
 [2]
 [6]
 [7]
 [0]]
```

**Comment by asmorkalov** [comment]
@dkurt Do you have progress with the issue?

**Comment by dkurt** [comment]
@asmorkalov not started yet

**Comment by asmorkalov** [comment]
Still reproducible with current 4.x (pre 4.13.0).
